### PR TITLE
move recursive efforts back into node methods, fix many-a-failing-test

### PR DESCRIPTION
**Pull request by alexthemayers**

### CODE REVIEW

File: openapi/node.go [hunk 1]
```diff
@@ -85,10 +100,19 @@ func (n *node) setName(name string) {
 }
 
 func (n *node) setRenderer(r Renderer) {
-	n.renderer = r
+	if n.parent == nil {
+		n.renderer = r
+		return
+	}
+	if n.parent.getRenderer() == nil {
```
Comment (alexthemayers) [inline openapi/node.go:107]: @alexthemayers possible bug

File: openapi/node.go [hunk 2]
```diff
@@ -36,12 +36,23 @@ const (
 )
 
 func (n *node) SetMetadata(metadata map[string]string) {
+	if n.parent.GetParent() == nil {
```
Comment (SasSwart) [inline openapi/node.go]: traversing 2 levels up is a smell. It should definitely be possible to stay within one level and that would be simpler.

File: openapi/node.go [hunk 3]
```diff
@@ -36,12 +36,23 @@ const (
...
+		root, ok := n.GetParent().(*OpenAPI)
```
Comment (SasSwart) [inline openapi/node.go]: Is `n.GetParent().(*OpenAPI)` realle necessary, or would `n.GetParent() == nil` be enough to tell you that you're at the top of the tree?

File: openapi/node.go [hunk 4]
```diff
@@ -63,9 +74,13 @@ func (n *node) setParent(parent Traversable) {
 
 // getBasePath recurses up the parental ladder until it's overridden by the *OpenAPI method
 func (n *node) getBasePath() string {
-	return n.parent.getBasePath()
+	if n.parent == nil {
```
Comment (SasSwart) [inline openapi/node.go:76]: nice

File: openapi/paths.go [hunk 5]
```diff
@@ -33,6 +33,9 @@ func (p *PathItem) GetPath() string {
 }
 
 func (p *PathItem) getBasePath() string {
+	if p.parent == nil {
```
Comment (SasSwart) [inline openapi/paths.go:36]: How do you feel about maintaining the stance that any type can be the root of the tree? Hardcoding the expectation of a specific type at the root makes the project inflexible. While I can't see an immediate need for anything other than *OpenAPI to be at the root, I do prefer to keep the extensibility.

Can we define a root node as any node that has `parent == nil`?

File: openapi/refs.go [hunk 6]
```diff
@@ -8,15 +8,16 @@ import (
 
 // readRef takes a reference and attempts to unmarshal it's content into the struct being passed as `i`.
 // As it happens, this ref is contained within the struct that is being unmarshalled into.
+// TODO can `i` become a non-empty interface? perhaps [Node | Traversable]?
 func readRef(absFilename string, i interface{}) error {
-	fmt.Printf("Reading reference: %s\n", absFilename)
+	if Debug { // this can be a particularly noisy Printf call
```
Comment (SasSwart) [inline openapi/refs.go]: Consider adding a cli flag to set Debug to true or false. If you feel that's beyond the scope of this PR, would you consider creating an issue for it?

File: openapi/schema.go [hunk 7]
```diff
@@ -63,3 +63,16 @@ func (s *Schema) getBasePath() string {
 func (s *Schema) getRef() string {
 	return s.Ref
 }
+
+func (s *Schema) IsRequired(property string) bool {
```
Comment (SasSwart) [inline openapi/schema.go:67]: Nice!

File: openapi/schema_test.go [hunk 8]
```diff
@@ -5,13 +5,13 @@ import (
 	"testing"
 )
 
-var (
-	properties = map[string]*Schema{
+func emptySchemaWith(childProperties, childItems, parent bool) *Schema {
```
Comment (SasSwart) [inline openapi/schema_test.go]: Consider functional option parameters here. They provide a less rigid way to initialise a Schema.

File: openapi/node.go (hunk 2 above)
Comment (alexthemayers) [reply openapi/node.go]: Yeah. In hindsight this now looks strange to me too.
Moved metadata to the top-most node, not OAS object, to fix this.

File: openapi/node.go (hunk 3 above)
Comment (alexthemayers) [reply openapi/node.go]: Metadata was previously directly attached to the OpenAPI object. I moved this into the node type/interface allowing for traversal as this allows access without making it public.
This fixed this issue as well as the previous one

File: openapi/refs.go (hunk 6 above)
Comment (alexthemayers) [reply openapi/refs.go]: Sure, I'll do that. 
Given the hope for adoption and quality of feedback we may receive through making that a public feature, I really like that idea

File: openapi/schema_test.go (hunk 8 above)
Comment (alexthemayers) [reply openapi/schema_test.go]: I'll give it a bash, but I'm unclear in how that might look

File: openapi/paths.go (hunk 5 above)
Comment (alexthemayers) [reply openapi/paths.go:36]: I would like to keep that definition, yes. 
Let's keep things extensible 💪

File: openapi/schema_test.go (hunk 8 above)
Comment (SasSwart) [reply openapi/schema_test.go]: No biggie.

**Review by SasSwart — APPROVED** [review]

**Pull request merged**